curriculum/challenges/english/07-scientific-computing-with-python/scientific-computing-with-python-projects/time-calculator.md

```diff
@@ -52,24 +46,382 @@ add_time("6:30 PM", "205:12")
 
 Do not import any Python libraries. Assume that the start times are valid times. The minutes in the duration time will be a whole number less than 60, but the hour can be any whole number.
 
-## Development
+# --hints--
+Calling `add_time()` with `3:30 PM, 2:12` should return `5:42 PM`.
+
+```js
+({
+  test: () => {
+    pyodide.FS.writeFile('/home/pyodide/time_calculator.py', code);
+    pyodide.FS.writeFile('/home/pyodide/test_module.py', `
+import unittest
+from time_calculator import add_time
 
-Write your code in `time_calculator.py`. For development, you can use `main.py` to test your `time_calculator()` function. Click the "run" button and `main.py` will run.
+class UnitTests(unittest.TestCase):
+    maxDiff = None
+    def test_same_period(self):
+        actual = add_time("3:30 PM", "2:12")
+        expected = "5:42 PM"
+        self.assertEqual(actual, expected, 'Expected calling "add_time()" with "3:30 PM", "2:12" to return "5:42 PM"')  
+`);
+    const testCode = `
+from unittest import main
+t = main(module='test_module', exit=False)
+t.result.wasSuccessful()
+`;
+    const out = __pyodide.runPython(testCode);
+    assert(out);
+  }
+})
```

Comment:
There's a bit funny issue, any solution can be used to show it.
1. Paste (write) solution, but change something that would not make it pass - ie. comment out `return` or change function signature to `def add_time(start, duration):`. In other words run tests on version that doesn't pass.
2. Run the Tests - they will fail, as expected.
3. Reverse the change.
4. Run the Tests - they will fail, for the same reason as in point 2. Not as expected.
5. Hit `F5`
6. Run the Tests - they will pass, as expected.

Generally python module imports are cached, so it looks like that's somehow causing interference either with how tests are run or with pyodide.

Here's ugly version, forcing module reload using python's `importlib.reload`, to reload `time_calculator` and `test_module` modules.

```
({
  test: () => {
    pyodide.FS.writeFile('/home/pyodide/time_calculator.py', code);
    pyodide.FS.writeFile('/home/pyodide/test_module.py', `
import unittest
import time_calculator

from importlib import reload

reload(time_calculator)


class UnitTests(unittest.TestCase):
    maxDiff = None
    def test_same_period(self):
        actual = time_calculator.add_time("3:30 PM", "2:12")
        expected = "5:42 PM"
        self.assertEqual(actual, expected, 'Expected calling "add_time()" with "3:30 PM", "2:12" to return "5:42 PM"')  
`);
    const testCode = `
from unittest import main
import test_module
from importlib import reload

reload(test_module)
t = main(module='test_module', exit=False)
t.result.wasSuccessful()
`;
    const out = __pyodide.runPython(testCode);
    assert(out);
  }
})
```

Reply:
That's exactly the conclusion I just came to. After quite some trial and error!

Basically, yes, pyodide is never reset, so the modules get 'stuck' in the cache. We do drop the globals between tests, but the module cache sticks around. Normally this is fine, because the modules don't change. However, when we're updating them inbetween tests, we get all kinds of weird stuff going on.

As far as I can see there are three ways out:

1) the ugly solution
2) rewrite them all without unittest or pytest
3) create one test that uses unittest/pytest as you would normally (i.e. a single file with one or more suites)

Option 3) could be quite nice, but we'd need to rework the test runner significantly to handle running a single JS test that calls a bunch of python tests. Probably too much work.

Option 2) should work just fine, but is also a fair chunk of work.

So, I think option 1) is the way to go.